integration/grid.py

```diff
@@ -0,0 +1,505 @@
+        yield deferLater(reactor, .1, lambda: None)
+    furl = _validate_furl(furl_fname)
+
+    returnValue(
+        Introducer(
+            process=TahoeProcess(transport, intro_dir),
+            protocol=protocol,
+            furl=furl,
+        )
+    )
+
+
+@attr.s
+class Grid(object):
+    """
+    Represents an entire Tahoe Grid setup
+
+    A Grid includes an Introducer, Flog Gatherer and some number of
+    Storage Servers.
+    """
+
+    _reactor = attr.ib()
+    _request = attr.ib()
+    _temp_dir = attr.ib()
+    _port_allocator = attr.ib()
+    introducer = attr.ib()
+    flog_gatherer = attr.ib()
+    storage_servers = attr.ib(factory=list)
+    clients = attr.ib(factory=dict)
+
+    @storage_servers.validator
+    def check(self, attribute, value):
+        for server in value:
+            if not isinstance(server, StorageServer):
+                raise ValueError(
+                    "storage_servers must be StorageServer"
+                )
+
+    @inlineCallbacks
+    def add_storage_node(self):
+        """
+        Creates a new storage node, returns a StorageServer instance
+        (which will already be added to our .storage_servers list)
+        """
+        port = yield self._port_allocator()
+        print("make {}".format(port))
+        name = 'node{}'.format(port)
+        web_port = 'tcp:{}:interface=localhost'.format(port)
+        server = yield create_storage_server(
+            self._reactor,
+            self._request,
+            self._temp_dir,
+            self.introducer,
+            self.flog_gatherer,
+            name,
+            web_port,
+        )
+        self.storage_servers.append(server)
+        returnValue(server)
+
+    @inlineCallbacks
+    def add_client(self, name, needed=2, happy=3, total=4):
+        """
+        Create a new client node
+        """
+        port = yield self._port_allocator()
+        web_port = 'tcp:{}:interface=localhost'.format(port)
+        client = yield create_client(
+            self._reactor,
+            self._request,
+            self._temp_dir,
+            self.introducer,
+            self.flog_gatherer,
+            name,
+            web_port,
+            needed=needed,
+            happy=happy,
+            total=total,
+        )
+        self.clients[name] = client
+        yield await_client_ready(client.process)
+        returnValue(client)
+
+
+
+# XXX THINK can we tie a whole *grid* to a single request? (I think
+# that's all that makes sense)
+@inlineCallbacks
+def create_grid(reactor, request, temp_dir, flog_gatherer, port_allocator):
+    """
+    """
+    intro_port = yield port_allocator()
+    introducer = yield create_introducer(reactor, request, temp_dir, flog_gatherer, intro_port)
+    grid = Grid(
+        reactor,
+        request,
+        temp_dir,
+        port_allocator,
+        introducer,
+        flog_gatherer,
+    )
+    returnValue(grid)
+
+
+def create_port_allocator(start_port):
```

Comment:
We have multiple existing port allocators already, might be nice to reuse an existing one if possible?

Reply:
in particular, it would be great to use the one that doesn't have the collided-with-another-listener behavior

Reply:
this might pre-date that in any case, but which one is that?

Reply:
The unit tests use `SameProcessStreamEndpointAssigner` / `AdoptedServerPort`.  The integration tests do allocation and bind in different processes?  So `SameProcessStreamEndpointAssigner` might need to be extended to expose the file descriptor directly so it can be passed down to the child process (which Twisted won't help us with on Windows ...).

Reply:
Right, the integration tests run things as a "real" user would, and Tahoe doesn't have a way to pass in already-bound ports.

This would require at least a couple tickets, including a "how does the configuration look" piece (and, as you point out, doesn't help even the integration tests on Windows). I do agree there's a race-window in there when this could conceivably fail -- but also don't know how a "normal" user might want to use such a feature? Perhaps Gridsync..?

Reply:
I thought that the integration tests would be able to use the "adopt-socket:" string endpoint description plugin to do this.  If that doesn't work (which maybe it doesn't ... because Tubs or something?) then a better path is probably to flip the interaction around: let Tahoe allocate the port but _expose_ its choice so that the other side (integration tests, GridSync, whatever) can discover the value.  This avoids weird socket passing shenanigans and so works on Windows too.

Reply:
..and Gridsync does something approximating that already I believe (does it just scrape the logs for the ports?) ... the downside being that you then you have to "run the software" before you learn any of the ports. Might still work here that way around..

Reply:
> the downside being that you then you have to "run the software" before you learn any of the ports

I think this only partially captures the picture.  You _always_ have to run the software before you learn any of the ports.  Before you run it, you don't know if it will be able to bind any of them or not.

Reply:
Sure ... but what I actually mean above is that "before anything else can be properly configured, you have to run the software" (which potentially means running it twice or expressing dependencies properly .. we already have some of this, e.g. to figure out what "the Introducer fURL is" we need to run it).

An additional approach is suggested in the comments: sent up CI so the OS is told to not auto-allocate a certain port-range (e.g. 45000+) and use ports in that range for the tests.

Then one knows that a port isn't bound (or, if it is, an error is appropriate).

Even without the above, if such expectations fail then we do learn about it (but in the bad way of "some tests fail, because the software failed to run").

Given that the existing tests _do_ use static ports, we can incrementally approach something better. I certainly don't want another 5-year long branch because it's not perfect yet (note however, that I haven't yet looked into or tried `adopted-stream:` or whatever because other things have bit-rotted).

Does anything currently use the adopted-stream thing? Systemd...? (A quick grep doesn't reveal any tahoe code using it)

Reply:
Yeah I don't consider this a blocker.

Reply:
@itamarst did you meant `allocate_tcp_port` from `iputil.py`? (It's a re-export from foolscap)

We could use that, but it would make the ports allocated non-deterministic (and of course be a foolscap dependency).

----

In order to use `adopt-stream:` we'd have to make it a more-general feature I think (it's in test/common right now so won't exist as-is for use in a normal setup as far as I can see). It would also make any integration test that uses it unable to run on Windows...(perhaps this could be useful for systemd usage?)

Reply:
Okay, I switched to the Foolscap one. Minor disadvantages: random ports instead of sequential (maybe harder to debug?) and it uses `socket` directly not "Twisted APIs". Meh.

On the plus side, it's less code :) seems to work and is synchronous.

(I'm certainly not opposed to using `adopt-stream` where possible, but that's a ton more work and won't work on Windows anyway so if we want that please file a different ticket).